### DISCUSSION TITLE
Documentation Updates

### DISCUSSION THREAD

**Discussion by JulianOrteil** [ponoor/step-series-support#6] [Ideas]
Per @kanta and my discussions at the end of [this issue](https://github.com/ponoor/python-step-series/issues/3), we decided to move it here to better expose our ideas and put them in a proper location.

To summarize our discussions in the above issue thus far: I am developing Ponoor's `python-step-series` library. This library is designed to enable easy communication with the STEP400, STEP800, and other Step-series devices Ponoor may develop in the future using Python. The library needs documentation in order for new and current users to know how it is meant to be used. Also, as I assume, @kanta is not a native English speaker, I am offering to review his current English documentation to improve it.

Because both of these need to be done, I floated the possibility of hosting the documentation on https://readthedocs.org/, the de-facto standard hosting site for virtually every Python library instead of on his website. A highlight of RTD is it will automatically compile new documentation from reStructuredText (reST) files and from the docstrings found in Python code. This makes the usually extremely tedious task of documenting behavior and features practically effortless.

However, this also means all current documentation, tutorials, and explanations he currently has on his website should be transferred to RTD if we are to move forward. It would be improper and confusing for users to be directed to two locations for documentation.

**Comment by kanta** [comment]
@JulianOrteil Thank you for opening the thread and summarize! Also thanks for information about the RTD system, seems very good. 
As you said, the original docs should  be removed to avoid the confusion. 
Let me have one more question: Is it possible to generate multilingual docs from the embedded codes?

**Reply by JulianOrteil** [reply]
Yes it is! There are a couple of ways actually. Manually writing them (so essentially you'll have what you have now, two identical copies just in EN and JA) is what I recommend. Otherwise there's a service called Transiflex that will do it for you to translate between two languages automatically.

I haven't written multilingual docs before, so I haven't used Transiflex; but it certainly wouldn't hurt to try :)

**Reply by kanta** [reply]
Great, it's a good news. Ok, then we are ready to go for RTD! Once the documentation is migrated to it, I believe it will greatly reduce the maintenance effort in the future.

Yes, let's try Transiflex some then we decide the way.
I haven't worked with RTD/sphinx/reStructuredText, so I would like you to handle the start-up this process.
Big thanks!!

**Reply by JulianOrteil** [reply]
My pleasure!

Because we both already have documentation, I want to focus on finalizing a few aspects of the library (mainly test coverage) first and then we'll tackle the documentation.

As I write the docs, please do tell me if I misrepresent or have leave something out. I plan on basically doing a copy paste of what you already have, just improved and grammatically fixed. The content will remain the same.

**Comment by JulianOrteil** [comment]
@kanta You describe the maximum phase current for the boards in "Motor Selection" under "Power supply and motor". Is this the maximum cumulative draw (so, for example, 8 motors on the STEP800 cannot exceed 2 amps in total), or the maximum draw per motor?

Jules

**Reply by JulianOrteil** [reply]
And then, what is the maximum power rating for the screw terminal block? (76V @ ??? amps on the STEP400 for example)

**Reply by kanta** [reply]
@JulianOrteil These are the maximum current per motor. So the cumulative current could be 16A for STEP800 (2A x8), or 20A for STEP400 (5A x4). 
STEP400 and STEP800 use a same screw terminal, [WJ25C-B-7.62-2P](https://lcsc.com/product-detail/Barrier-Terminal-Blocks_Ningbo-Kangnex-Elec-WJ25C-B-7-62-2P_C8422.html) from Ningbo Kangnex. According to it's datasheet, the maximum current is 24A.

> 24A -40℃~+105℃ 250V

**Reply by JulianOrteil** [reply]
Perfect, thank you for the info.

Also, we'll need to create an account with https://www.readthedocs.org. I recommend linking the ponoor organization account (or yours, whichever you prefer) here on Github for easy setup. Please follow the short tutorial [here](https://docs.readthedocs.io/en/stable/tutorial/#sign-up-for-read-the-docs) and choose `python-step-series` as the project to import, not the `rtd-tutorial` project they reference.

For now, I've set up a page for you to review my progress: https://jo-python-step-series.readthedocs.io/en/latest/step-series.html

**Reply by kanta** [reply]
Thank you for the direction!

I made an account on RTD and try to import  but it seems not possible to import a repo to multiple RTD projects. 
Probably better to share the RTD account with you?

![image](https://user-images.githubusercontent.com/1097902/150889284-696ff95c-e3b5-41a4-9a2d-99778ab2f1df.png)

BTW, the generated pages have much clearer text and design than originals. Very nice, thank you!

**Reply by JulianOrteil** [reply]
Shoot! I must have accidentally forgot to change the name of the project while testing my configuration. Unfortunately I deleted the repository, so if I am unable to recover access to the site, I will reach out to the RTD team for help.

Apologies.

**Reply by JulianOrteil** [reply]
Excellent. I was able to recover the project without issue. Please tell me the account username you have so O can add you as a maintainer.

Glad you like the updated docs though! I've racked my head over how to organize them (as I can't easily mimic your current set up) so if there's any changes you would like, please let me know.

Jules

**Reply by kanta** [reply]
My account on RTD is here https://readthedocs.org/profiles/kanta/

**Reply by kanta** [reply]
Oh, I'm sorry, is it the Github account you need?  https://github.com/kanta
And yes, as you said, how to organize the whole documentation is a challenging task. 
Ideally, the section "Running Basic Commands" would be easier to understand if it were explained for each environment (Python, TD, Max, etc...). But that's not our top priority, so let's get the whole thing going first.

**Reply by JulianOrteil** [reply]
No, just the name of your RTD account, thank you. You are now a maintainer and have full administrative access to the docs.

But I would argue that for Running Basic Commands, explaining how to setup the environment for Max, TD, etc is beyond the scope of the project. Kind of like how in Verify Your Connection, we link to external help on how set a PC's IP address depending on the OS, I would advocate for doing something similar for the programs.

Python, in my opinion, should be the only environment we should explain since the docs are centered around the library.

As always, let me know what you think.

Jules

**Reply by kanta** [reply]
Thank you for the setting up, now I can see the project from my account.

I thought your tutorial with Python was very clear and it would be nice to be able to prepare something similar in another environment, but in any case these are extra things (and I have no intention of asking you to write them!)

Then I organized my own thoughts as follows...
The documentation we are working on now is supposed to take over the role of the original documentation which will be erased in the future. Therefore, in my opinion, the docs needs an equality among the environments. For example. if the docs are too focused on the Python, it might be not friendly for non coders like [Vezer](https://imimot.com/vezer/) or [QLab](https://qlab.app/) users.
In that sense, how about describe the tutorial with OSC commands only, then prepare a link to a detailed explanation page with Python? Other environments can have a same structure in the future.

I understand that this is outside the scope of the Python library project. Once we establish a RTD workflow, I can work on porting the documentation so that you can focus on the review.
Let me know your thoughts, and thanks again for your big contribution!

**Reply by JulianOrteil** [reply]
Ahhhh, I see. I will leave that to you then. You shouldn't need any setup other than cloning the python library and then pushing doc changes as you would normally with code changes. RTD will see that change and automatically start building a new version of the docs for you. No Python setup should be necessary.

As for your idea on linking to detailed explanations for the library, I'm unsure if they would be necessary. The way I've designed the library is intended to make it extremely easy to use.

For example, these 4 lines of code are all that are needed to establish communication with the device:
```python
from stepseries import commands
from stepseries.step400 import STEP400

device = STEP400(1, "10.0.0.101")
device.set(commands.SetDestIP())
```

And then handling a command like `/getVersion` would be:
```python
# Additional import
from stepseries import responses

def version_handler(msg):
    print(msg)

# Register a handler to only handle responses of /version
# All other responses are ignored by this handler
device.on(responses.Version, version_handler)

# Send the get command
# Note the response is also returned here as well
msg = device.get(commands.GetVersion())
```

I would essentially note: each command is a template, they are either get commands or set commands, register your callbacks and that would be it. Then, in the docstring for each command, I would break down, like how you have it in your docs, the types and valid values for each parameter, a description and the response they can expect. I'd provide, obviously, examples for some commands, but it shouldn't be necessary for all of them.

I hope that clarifies my current intentions. Let me know if you would like me to take it another way.
Jules

**Reply by JulianOrteil** [reply]
Perhaps I missed this, but do you believe it would be best to replace the "Running Basic Commands" body in `STEP400/STEP800` with a list of links to dedicated tutorials for each program/environment that would then essentially contain the current body (just with dedicated screenshots/explanations pertaining to that specific environment)?

For example, the Max tutorial would have screenshots or directions on what to click and where to type whereas the Python tutorial would be dedicated to code?

**Reply by kanta** [reply]
Thank you for the info for RTD, I will try it!

Perhaps my concern with "Running Basic Commands" is that, for non-Python users, it is difficult to find the OSC message that should be sent. To rotate a motor, it will be
`device.set(commands.Run(1, 200))`
in Python but for OSC,
`/run 1 200`

So sometimes users have to find these OSC messages from comments in the Python codes and I feel it's not easily understandable for non-coders. 

Perhaps this could be a small thing in the visual design. For example, instead of showing the OSC message as a comment in the Python code, how about expressing like this?

> Let's start running motor. Send this OSC message.
> `/run 1 200`
> For python-step-series,
> `device.set(commands.Run(1, 200))`

Or I can prepare an OSC only version page separately.

I think I understand that this Python library is really great to be able to easily manipulate the STEP series without referring to the documentation every time. At the same time, I want to make the documents easy to understand for total non-programmers, for example, who can only connect boxes on Max like my old self.

**Reply by JulianOrteil** [reply]
I have modified the [page](https://python-step-series.readthedocs.io/en/latest/step-series.html#running-basic-commands) to more explicitly differentiate between raw commands and Python code. When you start working on adding documentation for the third-party programs, we should then add a list of links below the section header to how to send commands using those programs. Lmk what you think

**Reply by JulianOrteil** [reply]
I have updated the docs again to show my approach to providing the documentation as seen in your OSC Commands Reference. I've done both `SetKval` and `GetKval`: https://python-step-series.readthedocs.io/en/latest/index.html

**Reply by kanta** [reply]
Thank you for updates! 
I think the API reference is beautifully organized and good. I am new to Python so I think what you find good is good for Python programmers.
Just to clarify, This Python library API reference exists separately from the OSC command reference, correct?

**Reply by JulianOrteil** [reply]
That is up to you. Personally, I think if we combined them, it would be best. It avoids duplicating work.

I updated the examples on the index page above to now include the raw OSC command syntax like the yellow highlights in your OSC Command Reference. I intend to follow those examples if you approve.

**Reply by kanta** [reply]
Thank you for updates.
After looking at updated pages, I was thinking again that we need to strictly separate the main body of the documents from Python.
The tutorial page should only contain OSC commands, and the explanation with Python code should be on a separate page. Also, the OSC command reference should be separate from the python API reference. The overlap will be very large, but for people who want to use OSC, the python API reference is harder to read than the current documentation.

I apologize for the persistent discussion. It is important to note that this document is about the entire step-series and not just the Python Library. Also I want to make it clear that step-series can be used even if the user cannot read/write Python.

I think one option is that we use the RTD based documents for the API reference of the Python library, and we continue to use the current documents on ponoor.com for rest of the Python API.
I like the fancy documentation with RTD but our goals are a bit different and may be difficult to combine. How do you think?

**Comment by kanta** [comment]
By the way I made a draft for `functional-description.rst` but I met this error when I tried to commit it. I'm trying to figure out what's going on. Do you have any idea?
Thanks!

```
Traceback (most recent call last):
  File "C:\Users\kanta\.conda\envs\python-step-series\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\kanta\.conda\envs\python-step-series\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\kanta\.conda\envs\python-step-series\lib\site-packages\pre_commit\__main__.py", line 1, in <module>
    from pre_commit.main import main
  File "C:\Users\kanta\.conda\envs\python-step-series\lib\site-packages\pre_commit\main.py", line 13, in <module>
    from pre_commit.commands.autoupdate import autoupdate
  File "C:\Users\kanta\.conda\envs\python-step-series\lib\site-packages\pre_commit\commands\autoupdate.py", line 20, in <module>
    from pre_commit.store import Store
  File "C:\Users\kanta\.conda\envs\python-step-series\lib\site-packages\pre_commit\store.py", line 4, in <module>
    import sqlite3
  File "C:\Users\kanta\.conda\envs\python-step-series\lib\sqlite3\__init__.py", line 23, in <module>
    from sqlite3.dbapi2 import *
  File "C:\Users\kanta\.conda\envs\python-step-series\lib\sqlite3\dbapi2.py", line 27, in <module>
    from _sqlite3 import *
ImportError: DLL load failed while importing _sqlite3: �w�肳�ꂽ���W���[����������܂���B

**Reply by JulianOrteil** [reply]
I have not seen that error before! It appears to be related to `pre-commit`, a popular tool I use to enforce certain coding practices between developers. You should be able to run `pip install pre-commit` to fix the issue, or if you prefer, you can remove the pre-commit hook from `python-step-series\.git\hooks` (I think on windows it is `pre-commit.exe`, just remove that single file).

Let me know if you encounter further issues.

**Reply by kanta** [reply]
Thanks! `pip install pre-commit` doesn't work (only output a list of `Requirement already satisfied:...` ) but removing `pre-commit` from `.git\hooks` works! Thank  you for your assistance!

**Comment by JulianOrteil** [comment]
Constant discussion only allows for more ideas, so don't worry about it! But this documentation is yours, so I will always defer to your judgment on how to organize it.

I'll plan on following your direction and keep the reference in the code specific to Python and then build out a separate reference like how you have it in your website. Ill also update the title of the documentation.

Thanks!
Jules
________________________________

**Reply by kanta** [reply]
Thanks! I believe this is the best way though not so systematic.
As you say, it would be great if we could make two references, one for Python libraries and one for OSC.
Will the work of porting the current documentation to RTD continue to help you?

**Reply by JulianOrteil** [reply]
It does, yes, thank you.

As a sidenote, I will be away for the next few days, so I won't be able to work on this until I get back (probably around Tuesday or Wednesday next week). Please let me know if you need anything while I am away and I'll do my best to get back to you in a reasonable amount of time.

**Comment by JulianOrteil** [comment]
@kanta I have ported more documentation from your website directly into RTD. Please review:
[SetDestIP](https://python-step-series.readthedocs.io/en/documentation/api/stepseries.html#stepseries.commands.SetDestIP), [GetVersion](https://python-step-series.readthedocs.io/en/documentation/api/stepseries.html#stepseries.commands.GetVersion), [GetConfigName](https://python-step-series.readthedocs.io/en/documentation/api/stepseries.html#stepseries.commands.GetConfigName), [ReportError](https://python-step-series.readthedocs.io/en/documentation/api/stepseries.html#stepseries.commands.ReportError), and [ResetDevice](https://python-step-series.readthedocs.io/en/documentation/api/stepseries.html#stepseries.commands.ResetDevice) and tell me if you approve of the formatting I have chosen. Some of the documentation provided is auto-generated and not of my own writing (like `Bases: stepseries.commands.OSCSetCommand`). Once you approve, I will finish off the Python portion of the documentation. I may request your approval for the ``responses`` side, but that will come later.

Thanks!
Jules

**Reply by JulianOrteil** [reply]
I have removed the ``Bases`` portion from the documentation. It is unneeded IMO.

**Reply by kanta** [reply]
@JulianOrteil Thank you for the formats plan! I think they are nicely organized and fine.
At the same time, I am too novice in Python to evaluate the documentation format. To be honest, I did not understand the meaning of each items. Probably I need to study more about the structure of a Python code and libraries (class, module, etc...). Anyway I am so sure that what you find good is good , so please go a head with your style!

**Reply by JulianOrteil** [reply]
My apologies @kanta. Git is having a fit right now on my end. I can't seem to get RTD to load my latest changes because of it. I meant to show you a much more refined version of the docs than what is there now. I'll update this if I can get this fixed.

**Reply by JulianOrteil** [reply]
@kanta Okay, I believe I have fixed the issue. I will implement more documentation throughout the day today.

**Reply by JulianOrteil** [reply]
@kanta Apologies for all of these pings. I have finished off documentation up to "Motor driver settings". If I have the opportunity, I may work on more documentation this weekend. I do need your help with clarifying the differences between ``/getStatus`` and ``/getConfigRegister``. To me, based on your descriptions, they appear to be virtually identical.

I would also like your help in interpretting the table in [/getStatus](https://ponoor.com/en/docs/step-series/osc-command-reference/motor-driver-settings/#getstatus_intmotorid). I am not sure how to determine what in that table is being sent based on the values in ``/status``.

**Reply by kanta** [reply]
@JulianOrteil Ok I see, the RTD sites still looks same at my end even with an another PC. I will wait more for the changes to take effect.

`/getStatus` and `/getConfigRegister` return a value of `STATUS` and `CONFIG` register of the motor driver respectively. 

`STATUS` is a 16bit length register and several information are implemented in each bit positions. Names of the bit and it's position are similar but different between STEP400(PowerSTEP01) and STEP800(L6470). Here are screenshots of the bit position table of [PowerSTEP01](https://user-images.githubusercontent.com/1097902/158019614-f3556b95-d8ca-4a18-8b4c-a36c485134a2.png) and
[L6470](https://user-images.githubusercontent.com/1097902/158019843-b95c2e7e-145d-47f2-a2d6-751a8816790d.png).

`CONFIG` is also slightly different between these two motor drivers, and most part of them are related with hardcoded circuit board design, like a clock source of the chip, PWM frequency generator, or etc...

So basically `/getStatus` and `/getConfigRegister` are only useful when you modified the firmware and debugging it. One option is making them as "Unofficial" or "Advanced" command, then separate them from the main docs.

Thank you for your big effort for documentation things!!! And let me know if you have any question.

**Reply by JulianOrteil** [reply]
@kanta Thank you, I appreciate the explanations. I was planning to provide a bit more descriptive behavior for the commands as, looking at them from a newbie point of view, I get confused as to what they are telling me and how they differ from each other. I'll note their advanced usage along with any others you think would also fit the category.

**Reply by JulianOrteil** [reply]
@kanta I have finished off the ``commands`` documentation for the library. You can review them here: https://python-step-series.readthedocs.io/en/documentation/module-reference.html#module-stepseries.commands

Please note that this is on a separate branch of the library, so the bottom left (might be lower-right) of your screen must look like the following to see the latest changes (if it doesn't, click on this and choose "documentation").
![image](https://user-images.githubusercontent.com/23083317/158733741-fa5f86ca-804d-43d6-874b-5a336d091b9c.png)

Please let me know of any commands you would like to list as "advanced" or "debugging only" so I can note that in the docs. I'll be moving onto the responses module and then cleaning up the rest of the library. Testing documentation will also be fixed up and made more user-friendly as well.

As always, let me know of any changes you would like made.

Thanks!
Jules

**Reply by kanta** [reply]
@JulianOrteil Thank you so much for your hard and extremely fast work! The new docs looks really organized and good for readability. 
Currently I have things to do but will finish in this weekend. I will start to review the command documentation in the next week. 

Thanks a lot!
kanta

**Reply by kanta** [reply]
@JulianOrteil I have read the commands documentation. I think the following commands might be better labeled as differently, since they are only useful when you are debugging the firmware or hardware.
- [GetAdcVal](https://python-step-series.readthedocs.io/en/documentation/module-reference.html#stepseries.commands.GetAdcVal)
- [GetStatus](https://python-step-series.readthedocs.io/en/documentation/module-reference.html#stepseries.commands.GetStatus)
- [GetConfigRegister](https://python-step-series.readthedocs.io/en/documentation/module-reference.html#stepseries.commands.GetConfigRegister)
- [ResetMotorDriver](https://python-step-series.readthedocs.io/en/documentation/module-reference.html#stepseries.commands.ResetMotorDriver)

Each explanation of commands are very clear, thank you!!

**Reply by JulianOrteil** [reply]
@kanta I'm glad you like them.

Those commands have been updated. I will start porting over the responses when I have the opportunity in the coming days.

I noticed you have been using the library for your own debugging and testing purposes. How has it been working for you?

**Reply by kanta** [reply]
Yes, your test code helped me a lot! As I wrote [here](https://github.com/ponoor/step-series-universal-firmware), I was trying to combine the firmware of STEP400, STEP800, and their several prototypes. Eventually there are 8 variations so far and it's impossible to check the behavior of each commands for each variants by hand. 
I use your test code to see that each boards respond correctly, and it worked!

**Comment by JulianOrteil** [comment]
@kanta I've been looking into using Transifex for automatically translating our English docs into Japanese. I do not think it will be possible for a few reasons:
1. Transifex is a subscription service.
2. Because of 1, we may lose our translations when the subscription runs out due to the conflicting workflows Transifex and manual translating require.

I recommend sticking with manual translations. Unfortunately, I cannot help with translating because I don't know a single word in Japanese. However, I can set the environment up for the process, but you'll have to do the heavy lifting.

Jules

**Reply by JulianOrteil** [reply]
Running the index page through Google Translate yields the following:

![image](https://user-images.githubusercontent.com/23083317/159836934-ab6167dc-bb28-48df-bdf0-aa93db50e7fb.png)

Again, as I don't know Japanese, these might read fine or they might read like a 5-year-old wrote them and I could never tell haha.

**Reply by kanta** [reply]
@JulianOrteil Thanks for your research! I agree to stick with manual translating. We already have Japanese documentation so I can work for porting. As you expected, the result of Google translate is quite weird! So sticking with manual translating is good idea. 
But thanks for your experiment anyway!

**Reply by JulianOrteil** [reply]
@kanta Haha, sounds good. Ill set up the environment sometime today so you can work on this when you have the chance.

I'll update you when everything is ready. :)

**Reply by JulianOrteil** [reply]
@kanta The environment should be ready for you to start translating (please run `git pull`).

To translate, follow these steps:
1. The files (`.po`) to edit are in `docs/locales/ja/LC_MESSAGES`
2. Please ignore `api.po` and, optionally, `license.po`
3. In these `.po` files, `msgid` is the English text. Please put the Japanese version in `msgstr`. You'll notice that the English text is occasionally broken into multiple lines. I'll leave it up to you if you would like to try and mimic that with Japanese (if it's even possible haha).
5. Push your changes as normal, ReadTheDocs will take care of everything afterward.
6. You can view your changes here: https://python-step-series.readthedocs.io/ja/latest/index.html
7. You can switch between JA and EN by choosing the language shown in the bottom-left menu:
![image](https://user-images.githubusercontent.com/23083317/160148161-5983cae4-cb2d-4b86-89dd-3a1f79477032.png)
I've updated RTD to automatically show our current working branch when you switch for both JA and EN.

Making changes in the future will now require an extra step on top of the list above. When making **any** changes to the library in the future (doesn't matter what you do), please run `tox -e docs` to update the environment. tox will automatically add or update `.po` files to include any new English text that needs translating.

Please let me know if you have any questions. I'll do my best to clarify any confusion this causes.
Jules

**Reply by kanta** [reply]
@JulianOrteil Thank you for the setting up, I'm so impressed this systematic code + docs + translating structure... This is beautiful.
I will start to work soon! I found a new branch `documentation-localization`, should I push the update to it?

**Reply by JulianOrteil** [reply]
@kanta My pleasure! I've never done multi-lingual docs before, so I'm looking at this for the first time same as you ..and I agree, this is a lot easier than I was anticipating.

As for that branch, no. That was just a testing branch for me to learn how to set everything up before implementing it for real in the `documentation` branch. Use the `documentation` branch.

I should hopefully have time to review your new English docs this weekend. Please let me review them before you start translating.

Jules
